cms/models/placeholdermodel.py

```diff
@@ -429,5 +431,48 @@ def mark_as_dirty(self, language, clear_cache=True):
         elif attached_model is StaticPlaceholder:
             StaticPlaceholder.objects.filter(draft=self).update(dirty=True)
 
+    def get_vary_cache_on(self, request):
+        """
+        Returns a list of VARY headers.
+        """
+        def inner_plugin_iterator(lang):
+            """See note in get_cache_expiration.inner_plugin_iterator()."""
+            if hasattr(self, '_all_plugins_cache'):
+                for instance in self._all_plugins_cache:
+                    plugin = instance.get_plugin_class_instance()
+                    yield instance, plugin
+            else:
+                for plugin_item in self.get_plugins(lang):
+                    yield plugin_item.get_plugin_instance()
```

Comment:
This gets all the information it needs from the existing _all_plugins_cache. It is, in fact, query-free to do this in my tests (and the unit tests). Only if _all_plugins_cache isn't available, will it "fallback" to the case that requires queries. I haven't seen this case happen yet, in my testing, but if it does, I've got it covered.